File: ruby/lib/br_validator/cnpj.rb
# frozen_string_literal: true

module BrValidator
  # Validates, normalizes, and formats Brazilian CNPJ numbers.
  #
  # Covers both the traditional numeric format and the alphanumeric format
  # introduced by Receita Federal.
  module Cnpj
    module_function

    ALLOWED_CHARS = %r{\A[A-Za-z0-9.\-/\s]+\z}.freeze
    FORMATTING_CHARS = %r{[.\-/\s]}.freeze
    NORMALIZED_SHAPE = /\A[A-Z0-9]{12}\d{2}\z/.freeze
    private_constant :ALLOWED_CHARS, :FORMATTING_CHARS, :NORMALIZED_SHAPE

    # Strips formatting and uppercases value.
    #
    # A digit-only strip would destroy alphanumeric CNPJ values, so only
    # separator characters are removed.
    def normalize(value)
      value.gsub(FORMATTING_CHARS, "").upcase
    end

    def character_value(char)
      char.ord - 48
    end
    private_class_method :character_value

    def calculate_check_digit(value)
      sum = 0
      weight = value.length == 12 ? 5 : 6

      value.each_char do |char|
        sum += character_value(char) * weight
        weight -= 1
        weight = 9 if weight == 1
      end

      remainder = sum % 11

      return 0 if [0, 1].include?(remainder)

      11 - remainder
    end
    private_class_method :calculate_check_digit

    # Returns true if value is a structurally and check-digit valid CNPJ,
    # numeric or alphanumeric.
    #
    # Accepts raw or formatted input but rejects unexpected characters.
    def is_valid?(value)
      return false unless value.match?(ALLOWED_CHARS)

      cnpj = normalize(value)

      return false unless cnpj.match?(NORMALIZED_SHAPE)

      first_digit = calculate_check_digit(cnpj[0...12])
      return false unless first_digit == cnpj[12].to_i

      second_digit = calculate_check_digit(cnpj[0...13])

      second_digit == cnpj[13].to_i
    end

    # Returns value as XX.XXX.XXX/XXXX-XX.
    #
    # If the normalized value has an invalid length, value is returned
    # unchanged.
    def format(value)
      cnpj = normalize(value)

      return value unless cnpj.length == 14

      "#{cnpj[0...2]}.#{cnpj[2...5]}.#{cnpj[5...8]}/#{cnpj[8...12]}-#{cnpj[12...14]}"
    end
  end
end

File: python/src/br_validator/cnpj.py
"""Validates, normalizes, and formats Brazilian CNPJ numbers.

Covers both the traditional numeric format and the alphanumeric format
introduced by Receita Federal.
"""

import re

_ALLOWED_CHARS = re.compile(r"^[A-Za-z0-9.\-/\s]+$")
_FORMATTING_CHARS = re.compile(r"[.\-/\s]")
_NORMALIZED_SHAPE = re.compile(r"^[A-Z0-9]{12}\d{2}$")


def normalize(value: str) -> str:
    """Strip formatting and uppercase value.

    A digit-only strip would destroy alphanumeric CNPJ values, so only
    separator characters are removed.
    """
    return _FORMATTING_CHARS.sub("", value).upper()


def _character_value(char: str) -> int:
    return ord(char) - 48


def _calculate_check_digit(value: str) -> int:
    total = 0
    weight = 5 if len(value) == 12 else 6

    for char in value:
        total += _character_value(char) * weight
        weight -= 1

        if weight == 1:
            weight = 9

    remainder = total % 11

    if remainder in (0, 1):
        return 0

    return 11 - remainder


def is_valid(value: str) -> bool:
    """Return True if value is a structurally and check-digit valid CNPJ.

    Numeric or alphanumeric. Accepts raw or formatted input but rejects
    unexpected characters.
    """
    if not _ALLOWED_CHARS.match(value):
        return False

    cnpj = normalize(value)

    if not _NORMALIZED_SHAPE.match(cnpj):
        return False

    first_digit = _calculate_check_digit(cnpj[:12])
    if first_digit != int(cnpj[12]):
        return False

    second_digit = _calculate_check_digit(cnpj[:13])

    return second_digit == int(cnpj[13])


def format(value: str) -> str:
    """Return value as XX.XXX.XXX/XXXX-XX.

    If the normalized value has an invalid length, value is returned
    unchanged.
    """
    cnpj = normalize(value)

    if len(cnpj) != 14:
        return value

    return f"{cnpj[:2]}.{cnpj[2:5]}.{cnpj[5:8]}/{cnpj[8:12]}-{cnpj[12:14]}"
